### PR TITLE
libfetchers: fix race in the Git filesystem object sink

### PR DESCRIPTION
**Pull request by Mic92**
The sink built its tree structure from worker threads, after the blob bytes were written. Workers finish in any order, so for "foo" then "foo/bar" the "foo/bar" worker could run first and create "foo" as a directory instead of rejecting it as a file. Hence the flaky GitUtilsTest.sink_no_parent_dir.

Reserve each node synchronously, in order; workers only fill in the object ID afterwards.

## Context

---

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Review by xokdvium — COMMENTED** [review]
I did notice this bug yesterday that my test uncovered. Thanks for doing this! It occurs to me that this code is very buggy and lacks good test coverage currently.

Would be nice if we also had tests for all the corner cases:

- A non-directory parent `foo` when we write file/symlink/hardlink `foo/bar`. The test I added surfaced this particular case.
- Replacing a file/directory/symlink/hardlink with another type of node. What if we try to replace a non-empty directory? Should that succeed/fail?

I'm not sure which behavior that we want. Probably not exactly matching libarchive which is very hard to understand - exactly when stuff gets unlinked/rejected (when unpacking to a filesystem), but a bit stricter if possible. (Or maybe match the behaviour prior to parallel implementation to be on the safer side).

Also, ideally we wouldn't do this as adjusting weird semantics of a FileSystemObjectSink and had a separate interface for unpacking a tarball (a TarSink, cc @Ericson2314).

**Comment by xokdvium** [comment]
To be clear, this is certainly an improvement and untangling this mess probably should be done in a normal release cycle if possible, with proper tests and defined semantics. But fixing nondeterminism is a good start at least.

File: src/libfetchers/git-utils.cc [hunk 1]
```diff
@@ -1232,6 +1232,28 @@ struct GitFileSystemObjectSinkImpl final : GitFileSystemObjectSink
...
 
+    /* Set the object ID of a reserved leaf, skipping if it was superseded (id changed) meanwhile. */
+    void setNodeOid(State & state, const CanonPath & path, const git_oid & oid, size_t id)
+    {
+        auto parent = path.parent();
+        assert(parent);
+
+        Directory * cur = &state.root;
+        for (auto & name : *parent) {
+            auto i = cur->children.find(std::string(name));
```
Comment (xokdvium) [inline src/libfetchers/git-utils.cc]: Doesn't the map have a transparent comparator? (Aka `std::less<>` or something?)

File: src/libfetchers/git-utils.cc [hunk 2]
```diff
@@ -1232,6 +1232,28 @@ struct GitFileSystemObjectSinkImpl final : GitFileSystemObjectSink
...
+            if (i == cur->children.end())
+                return;
+            auto dir = std::get_if<Directory>(&i->second.file);
+            if (!dir)
+                return;
+            cur = dir;
+        }
+
+        auto i = cur->children.find(std::string(*path.baseName()));
```
Comment (xokdvium) [inline src/libfetchers/git-utils.cc]: Same here, should this work with a transparent comparator without creating a string?

**Comment by xokdvium** [comment]
Got started on cleaning this up, but discovered https://github.com/NixOS/nix/pull/16090 instead...

File: src/libfetchers/git-utils.cc (hunk 1 above)
Comment (Mic92) [reply src/libfetchers/git-utils.cc]: Done.

File: src/libfetchers/git-utils.cc (hunk 2 above)
Comment (Mic92) [reply src/libfetchers/git-utils.cc]: Done.

**Comment by xokdvium** [comment]
I'm going to put an alternative setup which gets rids of mutexes altogether soonish. Would simplify things a lot

**Review by edolstra — APPROVED** [review]

**Review by xokdvium — APPROVED** [review]
Guess we can clean up later.
